Change nan fallback in MOS Table to avoid JSON Serialization Warning

duytnguyendtn opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

Our traitlet messaging infrastructure does not support `nans` in the payload, as they are not JSON serializable. I changed the default fallback for missing RA/Dec values from `nan` to our default `Unspecified` we use for the Filter and Grating. Unfortunately this means it's a string rather than a float, and thereby a different type, but oh well.

I also add a preemptive sanitization check in `_add_to_table` to convert any `nans` to string 'nan's in case the user actually provides a `nan` directly. The issue/slightly controversial step here is the need to add `pandas` to our dependencies to actually perform the nan check. `numpy.isnan` doesn't support complex objects and will spit an error, as opposed to `pandas.isna`. We can debate whether we want this or not; I can easily remove it

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #<Issue Number>

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

rosteen reviewed (approved):
I'm not sure exactly how to test this, but it looks like it does the job.

pllim commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -50,6 +51,9 @@ def _add_to_table(app, data, comp_label, table_viewer_reference_name='table-view
         return
 
+    # Convert any nans to strings for JSON serialization
+    data = [str(x) if pd.isna(x) else x for x in data]
```
Seems a little overkill to introduce pandas as required dependency just to use `isna`.

duytnguyendtn replied:
That was my concern as well, but the only way to introduce this check that I could reasonably find. I wanted to gather thoughts as to whether this is something we want or not.

The plus is that pandas is already an indirect dependency 🤷

pllim replied:
You mentioned that `np.isnan` does not support complex object, but is this RA/Dec problem that complex?

pllim replied:
> pandas is already an indirect dependency

Huh, which packaging is pulling it in?

Regardless, before this patch, Jdaviz might still work if someone do a `pip uninstall pandas`, but that is no longer the case.

duytnguyendtn replied:
> is this RA/Dec problem that complex?

This isn't necessarily only for the RA/Dec; this is for any element that the user wants to add to the table. This came up because we add the actual Glue Data objects for each row to the table itself in hidden columns ("2D Spectrum", "1D Spectrum", "Image"). So `np.isnan` was failing when it was trying to inspect the Glue Data objects

duytnguyendtn replied:
Removed! See https://github.com/spacetelescope/jdaviz/pull/1958#issuecomment-1403934536

pllim commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -50,6 +51,9 @@ def _add_to_table(app, data, comp_label, table_viewer_reference_name='table-view
         return
 
+    # Convert any nans to strings for JSON serialization
+    data = [str(x) if pd.isna(x) else x for x in data]
```
Is there a reason why we cannot blindly pass in str representation of everything for serialization?

```suggestion
    data = [str(x) for x in data]
```

duytnguyendtn replied:
I think that depends on what is listening on the other side for the messages. If we change everything to strings, then we'll need to make sure to "recast" anything trying to use a message payload. Luckily for RA/Dec, most fields that would have "nans" would probably be string-compatible

pllim replied:
Whether who is listening I am not sure... I am just saying that if RA/Dec is the only thing tripping this check, pandas is probably unnecessary.

duytnguyendtn replied:
Removed! See https://github.com/spacetelescope/jdaviz/pull/1958#issuecomment-1403934536

duytnguyendtn commented:
As per discussion with the team, the included nan-check didn't seem worth the additional dependency, so I removed the check and rebased. This should be ready for rereview!

pllim reviewed (approved):
Simple and elegant. Thanks!

pllim commented:
Is a change log needed? I am on the fence.

pllim commented:
Ooops... looks like this affects tested behavior:

`ValueError: 'Unspecified' did not parse as unit: At col 0, Unspecified is not a valid unit.`

pllim commented:
Probably wanna add extra logic to `return None, None` in this check for your new values:

https://github.com/spacetelescope/jdaviz/blob/74c7655b3859fee406b4085def79948abe3f9fa9/jdaviz/configs/mosviz/helper.py#L335-L336

duytnguyendtn commented:
Tests are passing! (Just pushed up a changelog, but nothing in the code changed). Thanks for the assist @pllim!

pllim commented on CHANGES.rst:
```diff
@@ -87,6 +87,8 @@ Mosviz
 - Fix broken slit overlay plugin which resulted in a snackbar error being raised complaining about
   ``S_REGION``. [#1957]
 
+- RA Dec fallback values changed to "Unspecified" to avoid json serialization warning [#1958]
```
Is this accurate?

```suggestion
- RA/Dec fallback values changed to "Unspecified" to avoid JSON serialization warning when loading data. [#1958]
```

duytnguyendtn replied:
Works for me!

pllim commented:
You got 2 approvals, so feel free to merge when CI is green. 😸

duytnguyendtn commented:
Thanks for the eyes!

duytnguyendtn commented:
Grr... auto backport failed. Do we care enough to manually backport?
